### PR TITLE
remove some DATAPTR

### PR DESCRIPTION
**Pull request by jangorecki**
#3301, some obvious DATAPTR to remove
@mattdowle check one inline comment

### CODE REVIEW

File: src/uniqlist.c [hunk 1]
```diff
@@ -65,7 +65,7 @@ SEXP uniqlist(SEXP l, SEXP order)
       }
     } break;
     case STRSXP : {
-      SEXP *vd=(SEXP *)DATAPTR(v), prev, elem;   // TODO: tried to replace DATAPTR here but (SEXP *)&STRING_ELT(v,0) results in lvalue required as unary ‘&’ operand
+      SEXP *vd=(SEXP *)CHAR(v), prev, elem;
```
Comment (jangorecki) [inline src/uniqlist.c]: `CHAR` I found in http://biostat.mc.vanderbilt.edu/wiki/Main/RInternals please take a look before merging @mattdowle

File: src/fmelt.c [hunk 2]
```diff
@@ -618,7 +618,7 @@ SEXP getidcols(SEXP DT, SEXP dtnames, Rboolean verbose, struct processData *data
         // From assign.c's memcrecycle - only one SET_STRING_ELT per RHS item is needed to set generations (overhead)
         for (k=0; k<data->nrow; k++) SET_STRING_ELT(target, k, STRING_ELT(thiscol, k));
         for (j=1; j<data->lmax; j++)
-          memcpy((char *)DATAPTR(target)+j*data->nrow*size, (char *)DATAPTR(target), data->nrow*size);
+          memcpy((char *)CHAR(target)+j*data->nrow*size, (char *)CHAR(target), data->nrow*size);
```
Comment (jangorecki) [inline src/fmelt.c]: and here `CHAR` macro

Comment (jangorecki) [reply src/fmelt.c]: this should stress this change in `getidcols` fun, tested on master vs this branch and looks good
```r
d = data.table(a=as.character(letters[1:5]), v1=1:5, v2=1:5)
d[, "b" := paste0(a,rev(a),a)]
d[, "a" := paste0(a,a,a)]
melt.data.table(d, id=c("a","b"), measure.vars=c("v1","v2"))
```

File: src/rbindlist.c [hunk 3]
```diff
@@ -736,8 +736,8 @@ SEXP rbindlist(SEXP l, SEXP sexp_usenames, SEXP sexp_fill, SEXP idcol) {
       case INTSXP:
       case LGLSXP:
         if (TYPEOF(thiscol) != TYPEOF(target)) error("Internal logical error in rbindlist.c, type of 'thiscol' should have already been coerced to 'target'. please report to data.table issue tracker.");
-        memcpy((char *)DATAPTR(target) + ansloc * SIZEOF(thiscol),
-             (char *)DATAPTR(thiscol),
+        memcpy((char *)LOGICAL(target) + ansloc * SIZEOF(thiscol),
+             (char *)LOGICAL(thiscol),
```
Comment (mattdowle) [inline src/rbindlist.c]: We can't use LOGICAL on a REALSXP.  That fails with a check that R does when USE_RINTERNALS is not defined.

File: src/uniqlist.c (hunk 1 above)
Comment (mattdowle) [reply src/uniqlist.c]: See https://github.com/Rdatatable/data.table/pull/3350#pullrequestreview-199731517

File: src/rbindlist.c (hunk 3 above)
Comment (jangorecki) [reply src/rbindlist.c]: oops, I missed it refers to REAL, INT and LGL. Thanks

**Pull request merged**